### ISSUE TITLE
Move tracing to OpenTelemetry

### ISSUE THREAD

**Issue by bwplotka** [thanos-io/thanos#1972]
It would be nice to reuse ready libraries then invent and maintain our own (: 

AC:
* Our tracing library is replaced by https://opentelemetry.io/
* Support for existing providers is preserved: Jaeger, Elastic, Lightstep, Stackdriver

**Comment by squat** [comment]
Would love to see this. And we’ll be in line with other projects moving in this direction eg prom

**Comment by kakkoyun** [comment]
I'm happy to work on this.

**Comment by daixiang0** [comment]
@bwplotka i am happy to work on this.

**Comment by bwplotka** [comment]
Awesome! Thanks for taking a look. My thoughts:

* What about different backends (e.g lightstep, ELK APM) that OpenTelemetry does not support?
* Definitely IF we are doing the change, let's start small, so option 1.

> BTW, OpenTelemetry is still under active development, some features in OpenTracing may be not supported i think.

* Yes, that's my worry. That's why we should be careful. Let's focus now on what is missing that we are using? I guess small try does not hurt, if you want to give it a go - but you need to make sure you test it against e.g some open source provider like Jaeger.

This is kind of worrying state of Go right now. Maybe it's worth to put a friendly issue on https://github.com/open-telemetry/opentelemetry-go github with a question for current dates/blockers?

![image](https://user-images.githubusercontent.com/6950331/73429195-f4a37400-4332-11ea-81f1-bd9d9c0152e7.png)

**Comment by daixiang0** [comment]
I file a [issue](https://github.com/open-telemetry/opentelemetry-go/issues/453) about Elastic/Lightstep support, i would try the bridge.

**Comment by lawrencejones** [comment]
Given opencencus and opentracing are both mature libraries with LTS guarantees, would it make more sense to pick one of those for now? I've held off using opentelemetry while it's in alpha to avoid needless changes, and under the expectation that opencencus/opentracing -> opentelemetry will eventually become a well worn migration path.

**Comment by bwplotka** [comment]
We use opentracing right now, but it does not define actual RPC protocol, only client API. So we are forced to maintain client code against different tracing backends ourselves. Something that we would like to avoid

**Comment by kakkoyun** [comment]
As far as I know, OpenTelemetry has a fairly stable API for tracing. 
However, they are still in alpha, maybe we should wait for a little bit more. https://opentelemetry.io/project-status/

**Comment by daixiang0** [comment]
Yep, here is [the roadmap for opentelemetry](https://github.com/open-telemetry/opentelemetry-specification/blob/a131ae1ccabee8cc8f3b8cf109b81be378006fea/milestones.md), maybe we can wait for the 1.0 version of golang based on [feedback](https://github.com/open-telemetry/opentelemetry-go/issues/453#issuecomment-583215808).

**Comment by bwplotka** [comment]
Still todo.

**Comment by brancz** [comment]
We still want this.

**Comment by bwplotka** [comment]
Todo
Kind Regards,
Bartek Płotka (@bwplotka)

**Comment by kakkoyun** [comment]
Still valid.

**Comment by bwplotka** [comment]
On it, but missing Otel API -> opentracing tracer/exporters support that will allow us to use our custom exporter to Elastic APM/lightstep etc... .This is required to make sure we don't break compatibility of existing Tracing integrations when moving to Otel lib ):

**Comment by bwplotka** [comment]
https://github.com/open-telemetry/opentelemetry-go/issues/2193

**Comment by matej-g** [comment]
I had a quick look and it seems like the configuration should be translatable relatively good to ensure smooth transition from OpenTracing to OpenTelemetry (even if not completely 1:1). I think that would make more sense than having a 'translating' exporter (that would necessitate we still rely on OpenTracing packages, which I think we want to avoid, since the expectation is these will be deprecated).

To me it seems that:
1. For Jaeger we can switch to OpenTelemetry [Jaeger exporter](https://github.com/open-telemetry/opentelemetry-go/tree/main/exporters/jaeger) and this should be relatively smooth.
2. For GCP we can leverage the [Cloud Trace exporter](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/main/exporter/trace/README.md).
3. Other proprietary providers seems to support the OTLP format, which means we could leverage the [OTLP exporter](https://github.com/open-telemetry/opentelemetry-go/tree/main/exporters/otlp/otlptrace) and where I envision two possibilities:
a) We 'wrap' the proprietary tracer configurations into the OTLP exporter and use that to send spans.
b) We create a stand-alone OTLP tracer and keep the existing OpenTracing-based proprietary exporters with the understanding, that these will be deprecated and removed (after time) and users will be expected to use the OTLP exporter.
c) We'll do both a) and b)

I personally would be leaning towards option b). I'm wondering how many Thanos users are actually using these exporters. Since they are proprietary, I would expect the respective commercial providers to contribute the code and maintain it instead of community having to use resources to do that. Otherwise it makes more sense to have an OTLP exporter (which would be usable with any provider accepting it, seeing as it might become the standard and which is also accepted by the actual OTEL collector).

**Comment by bwplotka** [comment]
cc Cortex and Loki devs since they might laverage our traving clients. Are you ok with option `b` on this one? cc @pracucci @pstibrany @cyriltovena @slim-bean

**Comment by matej-g** [comment]
I'm posting to provide updates on the progress here. I started to migrate the exporters:
- We are currently limited to OTEL packages with version `v0.20.0` - this is due to the dependency chain, as we are importing (via Cortex and Grafana Dskit packages) module `go.etcd.io/etcd/client/v3` in version `v3.5.0`, which uses `v0.20.0` OTEL packages for its tracing. Due to changes in paths for some packages, I am not able to import latest OTEL version in Thanos since I'm getting `module * found, but does not contain package *`. The latest `master` branch of ETCD however imports newer OTEL version, so I'm expecting to be reconciled once this is released.
- For Jaeger, the OpenTracing client will soon be [sunset](https://github.com/jaegertracing/jaeger/issues/3362), however there are still
some features which are missing between the old client and the new OpenTelemetry exporter,
most notably we're missing remote sampler (which however is underway - see https://github.com/open-telemetry/opentelemetry-go-contrib/pull/936).
I have therefore postponed migrating this exporter until we can migrate with all features.
- I moved onto Google Cloud (Stackdriver) exporter as next, this one seems to be relatively simple to migrate, so good first candidate - opened a PR for this

I'll also use this comment to track the progress below:
- [ ] Migrate Jaeger client
- [ ] [Migrate Google Cloud (Stackdriver) client](https://github.com/thanos-io/thanos/pull/4838)
- [ ] Add OTLP exporter
- [ ] Migrate proprietary providers (ELM and Lighstep)
  - Alternative: (if cannot easily be migrated): Deprecate and point users to OTLP
- [ ] Migrate the instrumentation

Next step would be adding OTLP and once all necessary work to add Jaeger features on OTEL side is done, that will also be migrated.

**Comment by yeya24** [comment]
It would great if we can start with `Add OTLP exporter` step as this seems independent with other tasks.
If no one is working on this one right now then I can try to take it.

**Comment by matej-g** [comment]
Sorry for the delay on this @yeya24, I updated my [related PR](https://github.com/thanos-io/thanos/pull/4838) and commented there as well. I believe that PR should give us a common set of utilities to bridge OpenTelemetry exporters while we keep using OpenTracing instrumentation, for now. So I believe we can work on the other exporters independently once those bridging utilities land, as you say.

**Comment by bwplotka** [comment]
FYI: I played with much simpler API for own use https://github.com/bwplotka/tracing-go

**Comment by squat** [comment]
> FYI: I played with much simpler API for own use https://github.com/bwplotka/tracing-go (it using Otel underneath)

ooOOooh!

**Comment by Symbianx** [comment]
Looks like a PR for this issue was merged some time ago. When can we expect a new release with it?
